### PR TITLE
DataFormatDimMap supports 5D tensors and grappler layout optimizer supports ReduceOps for 5D tensors

### PR DESCRIPTION
**Pull request by kaixih**
This PR does:
(1) Allow DataFormatDimMap to support 5D dim tensors and map the dims to different combinations of NDHWC formats.
(2) Enable the grappler layout optimizer to handle ReduceOps when tensors are in 5D. This feature relies on (1), since the inputs of the reduce ops commonly contain a "reduce axis" which is a vector of dims. This "reduce axis" needs to be updated to correct format using DataFormatDimMap.

cc. @nluehr

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 1]
```diff
@@ -1371,18 +1371,45 @@ bool ReduceTransposer::IsReduceAxisSupported(
...
+  const auto& regular_fanin = node->GetRegularFanin(0);
+  const auto* output_shape_attr =
+      regular_fanin.node_view()->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
+  std::string dst_format = context->dst_format;
+  // Update the format from 4D to 5D layout if necessary.
+  if (rank == 5) {
+    std::string src_format_3d = src_format == "NHWC" ? "NDHWC" : "NCDHW";
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1383]: This seems very specific to CPU to GPU in regards to data formats. Can we check if the original src_formats and dst_formats are NHWC and NCHW only before applying this?

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc (hunk 1 above)
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1383]: I think that is correct. The layout opt only does the NHWC<->NCHW transpose. You can find that here that the src/dst formats are predefined: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc#L38

Comment (andyly) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1383]: If I remembered correctly the actual transposers can support other layouts. The optimizer though restricts the layouts, along with some GPU specific bits. I think it is a bit safer to only allow the 5d upgrade in the transposers if the layouts are of NHWC and NCHW for now (if the optimizer is updated to support other layouts).

Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1384]: Sure. I added one more barrier to restrict the optimizer only for NHWC/NCHW when 5d tensors are there.

**Review by andyly — APPROVED** [review]

File: tensorflow/python/grappler/layout_optimizer_test.py [hunk 2]
```diff
@@ -1194,6 +1197,42 @@ def testLeakyReluGrad(self):
...
+        metadata = config_pb2.RunMetadata()
+        output_val = sess.run(output, run_metadata=metadata)
+
+      nodes = []
+      num_transposes = 0
+      for node in metadata.cost_graph.node:
+        if _is_transpose(node.name):
+          num_transposes += 1
+        nodes.append(node.name)
+        print(node.name)
```
Comment (andyly) [inline tensorflow/python/grappler/layout_optimizer_test.py:1225]: Can this line be removed? This is triggering one of the format presubmits.

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/python/grappler/layout_optimizer_test.py (hunk 2 above)
Comment (kaixih) [reply tensorflow/python/grappler/layout_optimizer_test.py]: Sure. Removed this debugging print.

**Review by andyly — APPROVED** [review]

**Comment by andyly** [comment]
The tests between tf2xla and TF are shared, and the new test cases added for DataFormatDimMap are failing for the tf2xla kernel https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/data_format_ops.cc. Can this kernel also be updated?

File: tensorflow/python/grappler/layout_optimizer_test.py [hunk 3]
```diff
@@ -1194,6 +1197,41 @@ def testLeakyReluGrad(self):
...
       self.assertAllClose(output_val_ref, output_val, atol=1e-3)
 
+  @test_util.deprecated_graph_mode_only
+  def testReduceOpsFor5DTensors(self):
+    if test.is_gpu_available(cuda_only=True):
+      random_seed.set_random_seed(0)
+      x = random_ops.truncated_normal([1, 4, 2, 3, 3], seed=0)
+      w = random_ops.truncated_normal([2, 2, 2, 3, 3], seed=0)
+      gamma = random_ops.truncated_normal([1, 1, 1, 1, 3], seed=0)
+      beta = random_ops.truncated_normal([1, 1, 1, 1, 3], seed=0)
```
Comment (andyly) [inline tensorflow/python/grappler/layout_optimizer_test.py:1207]: I believe `gamma` and `beta` are unused? Can they be removed?

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/python/grappler/layout_optimizer_test.py (hunk 3 above)
Comment (kaixih) [reply tensorflow/python/grappler/layout_optimizer_test.py]: Sure. Removed.

**Comment by kaixih** [comment]
> The tests between tf2xla and TF are shared, and the new test cases added for DataFormatDimMap are failing for the tf2xla kernel https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/kernels/data_format_ops.cc. Can this kernel also be updated?

Sure. Done. PTAL.

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 4]
```diff
@@ -1371,18 +1371,46 @@ bool ReduceTransposer::IsReduceAxisSupported(
...
-  if (!ShouldProcess(*context, *node) || !IsFaninPortRankN(*node, 0, 4) ||
+  const auto& regular_fanin = node->GetRegularFanin(0);
+  const auto* output_shape_attr =
+      regular_fanin.node_view()->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
+  std::string dst_format = context->dst_format;
+  // Update the format from 4D to 5D layout if necessary.
+  bool allow_5d = rank == 5 && (src_format == "NHWC" || src_format == "NCHW");
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1382]: Sure. Done.

**Review by andyly — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

**Pull request merged**